### ISSUE TITLE
ATACorrect not work with different versions

### ISSUE THREAD

**Issue by HouyuZhang** [loosolab/TOBIAS#57]
Hi, developers:
Thanks for developing this excellent package for ATAC-seq footprinting. This package is powerful and seamless across different steps.   However, recently I found the `ATACorrect` not works when I update the version from `0.8.0` to `0.12.0` and `0.12.6`.  

In the `0.8.0` version, I successfully corrected my bam file and the log information suggested the post-bias variance reduced significantly than the pre-bias variance.
![image](https://user-images.githubusercontent.com/16659952/106349639-78a78780-630a-11eb-89d9-28c96c9b3796.png)
![image](https://user-images.githubusercontent.com/16659952/106349721-44809680-630b-11eb-9373-f5dc857ff083.png)

However, in the `0.12.0` and `0.12.6` version, it seems the correction failed, when I used exactly the same input file and parameter for the `ATACorrect`. I only show result from the `0.12.6` version here:
![image](https://user-images.githubusercontent.com/16659952/106349685-e784e080-630a-11eb-9845-849ce9eb2de7.png)
![image](https://user-images.githubusercontent.com/16659952/106349732-5a8e5700-630b-11eb-8075-ef0b8b84e462.png)

The command line I used is:
`TOBIAS ATACorrect --bam Human_K562_rep1.bam --genome Homo_sapiens.GRCh38.dna.primary_assembly.fa --peaks peaks.bed --blacklist hg38-blacklist.v2.bed --outdir ./ --cores 30`

I don't know whether the command lines or any parameters changed among these versions, which caused this inconsistency. Thanks in advance for your reply!  

Best Wishes,
Houyu.

**Comment by msbentsen** [comment]
Hi Houyu,

I tried to reproduce this error using the TOBIAS test data (from [here](https://github.com/loosolab/TOBIAS/wiki/test-data)) with versions 0.8.0 and 0.12.6, and I got the following results:

**0.8.0:**
```
2021-01-30 17:48:57 (46138) [STATS]	BIAS	pre-bias variance forward:	0.0035870
2021-01-30 17:48:57 (46138) [STATS]	BIAS	post-bias variance forward:	0.0000093
2021-01-30 17:48:58 (46138) [STATS]	BIAS	pre-bias variance reverse:	0.0036142
2021-01-30 17:48:58 (46138) [STATS]	BIAS	post-bias variance reverse:	0.0000085
```

**0.12.6:**
```
2021-01-30 17:55:14 (52209) [STATS]	BIAS	pre-bias variance forward:	0.0035870
2021-01-30 17:55:14 (52209) [STATS]	BIAS	post-bias variance forward:	0.0000138
2021-01-30 17:55:15 (52209) [STATS]	BIAS	pre-bias variance reverse:	0.0036142
2021-01-30 17:55:15 (52209) [STATS]	BIAS	post-bias variance reverse:	0.0000126
```

So there are some minor changes between these two versions ([CHANGES](https://github.com/loosolab/TOBIAS/blob/master/CHANGES) tells me that some things related to normalization were updated in 0.9.0) - but the bias correction seems to work in both cases. 

I am wondering if you had changes in other tools/dependencies between the two runs? Maybe that explains the differences? Because you are definitely right that the observed "pre-bias = post-bias" is an error!

Best regards,
Mette

**Comment by HouyuZhang** [comment]
Hi Mette,
Thanks very much for your reply! 
I downloaded the test dataset you provided and found a similar result as you listed above between `0.8.0` and `0.12.6`. This suggests your versions are working well.

After checking my preprocessing pipeline, I found I fed the shifted ATAC-seq reads (4, -5) into your `ATACorrect` procedure, besides, only the 5' or 3' which stands for Tn5 cut sites were fed (hereafter I will refer my input as `preprocessed`). When I used the raw bam file from Bowtie, the correction works on both versions ` 0.8.0` and `0.12.6`. 

**However, I still have several questions about these two versions:**
1. If I understand correctly, in the `ATACorrect` command, you will shift the reads (4, -5), and only the cut sites were reported in the `uncorrected.bw` and `corrected.bw`. Why directly input the `preprocessed` bam file works in version `0.8.0` but not `0.12.6`? 
Actually, even it works under `0.8.0` the output correction metric is weird because no PSSM information displayed and the Tn5 post-bias is not well as expected that a flat pattern will display (at least compared to your test dataset):
![image](https://user-images.githubusercontent.com/16659952/106374440-93d3cf00-63be-11eb-8c32-cf035e3206c4.png)
![image](https://user-images.githubusercontent.com/16659952/106374479-ba920580-63be-11eb-9c4a-b009d2e405f4.png)
I checked your CHANGES file and found you mentioned in the `0.9.0` version:

> Fixed rescale to original sum in ATACorrect

**Would you please explain in more detail what "rescaled" you made between these two versions?**

2. For the raw bam provided, both versions worked. However, the correction result differs slightly for both log information and the `_atacorrect.pdf` files:

`0.8.0`:
![image](https://user-images.githubusercontent.com/16659952/106374577-8c60f580-63bf-11eb-9998-feb592a86d40.png)
![image](https://user-images.githubusercontent.com/16659952/106374595-bb776700-63bf-11eb-93d5-44475eca2d27.png)

`0.12.6`:
![image](https://user-images.githubusercontent.com/16659952/106374615-d9dd6280-63bf-11eb-9d59-eedf3e848781.png)
![image](https://user-images.githubusercontent.com/16659952/106374606-c92cec80-63bf-11eb-9a82-d89f24e0a244.png)

It seems the correction by `0.8.0` is better, at least in my data. Again, would you please explain more details for both versions of the `ATACorrect` command, this will help me a lot.

**Another question, not about versions:**
1. I have carefully read your supplemental description of the `TOBIAS` algorithm in your paper (Bentse *et al*., 2020), you calculate the `expected` Tn5 cut sites using a total 100bp window flanking each location within the provided peak regions. I have tested the `--window` for 50 and 200 and found the results are largely similar, but I want to know why 100bp is the default value you set in the command? 

2. I have actually tested several packages for ATAC-seq footprinting before and found your package performs best and more easily manipulated. I found the main difference in your package is the Tn5 correction using DWM, which considers the di-nucleotide dependency. I wonder whether the tri-/ higher nucleotide dependency used will perform better or whether you have tested this? If higher nucleotide dependency metrics will be more computationally time-consuming?

Sorry for too many questions that bother you, I appreciate your further reply.
Best wishes,
Houyu.

**Comment by msbentsen** [comment]
Hi Houyu, your questions are not a problem! I will try to answer below:

>
>  If I understand correctly, in the `ATACorrect` command, you will shift the reads (4, -5), and only the cut sites were reported in the `uncorrected.bw` and `corrected.bw`. Why directly input the `preprocessed` bam file works in version `0.8.0` but not `0.12.6`?

I can't explain why it would work with one and not the other one - but I can say that ATACorrect expects the reads to be raw. If you have them shifted, you might set `--read_shift` to `0 0`, but I can't promise for the effects of that.

The difference in "rescale" between these two versions is, that the version `>0.9.0` rescales the corrected signal, so that **the sum of the corrected signal = sum of the original uncorrected signal** within each region. These signals were also correlated in value in version `0.8.0`, but the regions could lose signal strength if a high-bias-position was corrected. The correction stayed the same, but the code now just makes sure that the overall signal is comparable.

It might therefore be that the `0.8.0` looks better because certain regions were slightly under-represented. I would recommend using the higher version, as there might be other important bugfixes as well.
__________________
>I have tested the --window for 50 and 200 and found the results are largely similar, but I want to know why 100bp is the default value you set in the command?

I originally set the 100bp rather empirically based on previous literature on the subject, but also guess-work related to "how far away the effects of protein-binding and Tn5 bias is observed". If you see better effects with another window-size, feel free to use this!

>I wonder whether the tri-/ higher nucleotide dependency used will perform better or whether you have tested this? If higher nucleotide dependency metrics will be more computationally time-consuming?

I have not tried this, but increasing the nucleotide dependency will definitely be more computationally time-consuming. The step from PWM to DWM is already a lot slower even with cython optimization. In this case where Tn5 binds as a dimer, I imagine that the dinucleotide matrix is enough (as we are catching the dinucleotides between the two sub-motifs), but there might be cases where a tri+ nucleotide matrix is better. 

Best regards,
Mette

**Comment by HouyuZhang** [comment]
Hi Mette,
Thanks again for your detailed reply.

> I can't explain why it would work with one and not the other one - but I can say that ATACorrect expects the reads to be raw. If you have them shifted, you might set --read_shift to 0 0, but I can't promise for the effects of that.
Actually, when I conducting `ATACorrect` on my `preprocessed` bam file, I have set the --read_shift to 0 0.

**Comment by HouyuZhang** [comment]
Hi Mette,
Thanks again for your detailed reply.

> I can't explain why it would work with one and not the other one - but I can say that ATACorrect expects the reads to be raw. If you have them shifted, you might set --read_shift to 0 0, but I can't promise for the effects of that.

Actually, when I conducting `ATACorrect` on my `preprocessed` bam file, I have already set the `--read_shift` to `0` `0` for both versions `0.8.0` and `0.12.6`. So, I agree with you that this might not be a problem. Besides, I want to know whether all reads information were used for correction or only cut sites (5' or 3') were used. If only cut sites were used, I think it may explain why input the `processed` bam works in the version `0.8.0`. 

Thanks for the explanation on "rescale". If I understand correctly, the corrected signals measure the deviation between observed signals and expected signals at this position (Ci = Xi - Ei). The expected Tn5 insertion was estimated within a total 100-bp window flanking this position, with the underlining Tn5 DNA sequence bias considered (bi hat). In my understanding, a negative corrected value in the `corrected.bw` file indicates that the observed Tn5 signal in a position is smaller than expected and a positive corrected value indicates the observed Tn5 signal is larger than expected. 
So this is measuring **relative Tn5 signals instead of actual Tn5 signals**, and I guess this is why you need to rescale the corrected deviations to an equal sum of uncorrected signals. However, if I only want to use the relative Tn5 signals, and the original uncorrected signal is out of consideration in my analysis, will the version `0.8.0` ok? 

Thanks again for your kindly help!
Best wishes,
Houyu.

**Comment by msbentsen** [comment]
> 
> Besides, I want to know whether all reads information were used for correction or only cut sites (5' or 3') were used. If only cut sites were used, I think it may explain why input the `processed` bam works in the version `0.8.0`.

Only the 5' end of each read is used, meaning the start of forward reads, as well as start of reverse reads (on the opposite strand, so technically still 5'). But this was the case in both versions, so I am still a little confused about that...

>So this is measuring relative Tn5 signals instead of actual Tn5 signals, and I guess this is why you need to rescale the corrected deviations to an equal sum of uncorrected signals. However, if I only want to use the relative Tn5 signals, and the original uncorrected signal is out of consideration in my analysis, will the version 0.8.0 ok?

You can imagine a very small example like:
```
Observed:     1     0     0     4     1     2     1    (sum: 9)
Expected:     1     0     1     6     1     0     0    (sum: 9)
Corrected:    0     0    -1    -2     0     2     1    (sum of absolute values: 6   <- technically a loss of signal)
```
In version `>0.8`, the sum of corrected would be adjusted to be similar to the original observed cuts. So yes, if this is not needed for you, you can use 0.8. Although I will always recommend the newest version, I think both versions work well, and it is up to you to choose which one you find to be the best :-)

**Comment by HouyuZhang** [comment]
Thanks for your reply, Mette.

You have already perfectly answered all my questions, and I have no more questions now.
I will continue to use your excellent package in future studies. Thanks again for all!

Best wishes,
Houyu.

**Comment by HouyuZhang** [comment]
Hi, Mette:
Sorry for the delayed feedback, since I just back from vacation recently.
Thanks for your improvement on version `0.12.9`. This version works well on my `preprocessed` 1bp bam file, assessed from the log information and the `_atacorrect.pdf` file:
![image](https://user-images.githubusercontent.com/16659952/109385629-36389100-7930-11eb-8c05-2e07f87833c9.png)
![image](https://user-images.githubusercontent.com/16659952/109385643-4badbb00-7930-11eb-9b69-9292bf8dbaa0.png)
I wonder whether this bug exists in version `0.8.0` and if so, is this possible to fix this bug and build a new version from `0.8.0` to `0.8.1`. I will be much appreciative of this since I mainly used this version in my study.
Best wishes.
Houyu

**Comment by msbentsen** [comment]
Hi Houyu, 
I usually wouldn't provide support for earlier versions, and I would always recommend to use the latest version - but I will make an exception for polite requests like this :-) I just released 0.8.1 on [pypi](https://pypi.org/project/tobias/0.8.1/), so I hope that helps you out!

Best Mette

**Comment by HouyuZhang** [comment]
Hi, Mette,
Thank you for releasing the `0.8.1` version, it works well on my data, it seems the correction was much better than version `0.8.0`:
![image](https://user-images.githubusercontent.com/16659952/109651475-8a927980-7b99-11eb-88a4-2affc42026a9.png)
![image](https://user-images.githubusercontent.com/16659952/109651499-91b98780-7b99-11eb-817c-6ded3d27bfd6.png)

Thanks again for your time and help!
Best wishes,
Houyu.
